R/mock-session.R

```diff
@@ -0,0 +1,281 @@
+        return(NULL)
+
+      fileData <- readBin(file, 'raw', n=bytes)
+      b64 <- rawToBase64(fileData)
+      return(paste('data:', contentType, ';base64,', b64, sep=''))
+    },
+
+    setInputs = function(...) {
+      vals <- list(...)
+      mapply(names(vals), vals, FUN = function(name, value) {
+        private$.input$set(name, value)
+      })
+      private$flush()
+    },
+
+    .scheduleTask = function(millis, callback) {
+      id <- private$timer$schedule(millis, callback)
+
+      # Return a deregistration callback
+      function() {
+        invisible(private$timer$unschedule(id))
+      }
+    },
+    elapse = function(millis) {
+      msLeft <- millis
+
+      while (msLeft > 0){
+        t <- private$timer$timeToNextEvent()
+
+        if (is.infinite(t) || t <= 0 || msLeft < t){
+          # Either there's no good upcoming event or we can't make it to it in the allotted time.
+          break
+        }
+        msLeft <- msLeft - t
+        private$timer$elapse(t)
+
+        # timerCallbacks must run before flushReact.
+        private$timer$executeElapsed()
+        private$flush()
+      }
+
+      private$timer$elapse(msLeft)
+
+      # Run again in case our callbacks resulted in a scheduled
+      # function that needs executing.
+      private$timer$executeElapsed()
+      private$flush()
+    },
+
+    .now = function() {
+      # Contract is to return Sys.time, which is seconds, not millis.
+      private$timer$getElapsed()/1000
+    },
+
+    defineOutput = function(name, value, label) {
+      obs <- observe({
+        # We could just stash the promise, but we get an "unhandled promise error". This bypasses
+        prom <- NULL
+        tryCatch({
+          v <- value(self, name)
+          if (!promises::is.promise(v)){
+            # Make our sync value into a promise
+            prom <- promises::promise(function(resolve, reject){ resolve(v) })
+          } else {
+            prom <- v
+          }
+        }, error=function(e){
+          # Error running value()
+          prom <<- promises::promise(function(resolve, reject){ reject(e) })
+        })
+
+        private$outs[[name]]$promise <- hybrid_chain(
+          prom,
+          function(v){
+            list(val = v, err = NULL)
+          }, catch=function(e){
+            list(val = NULL, err = e)
+          })
+      })
+      private$outs[[name]] <- list(obs = obs, func = value, promise = NULL)
+    },
+
+    getOutput = function(name) {
+      # Unlike the real outputs, we're going to return the last value rather than the unevaluated function
+      if (is.null(private$outs[[name]]$promise)) {
+        stop("The test referenced an output that hasn't been defined yet: output$", name)
+      }
+      # Make promise return
+      v <- extract(private$outs[[name]]$promise)
+      if (!is.null(v$err)){
+        stop(v$err)
+      } else {
+        v$val
+      }
+    },
+
+    registerDataObj = function(name, data, filterFunc) {},
```

Comment:
Should the all the no-op functions give warnings?

Reply:
Perhaps. My mentality here was that there are some methods that you might expect to work (e.g. bookmarking), but there are others that are pretty meaningless in a mock session (e.g. sendCustomMessage). 

I didn't want to litter the console with a bunch of warnings for every module that uses sendCustomMessage when (hopefully) everybody already understands that there's no client to send a message to in this setup. Effectively my goal was to warn for any method that I think we might implement in the future until it's implemented.

But I could be convinced that it might be more consistent to always warn. The other thing I'd considered was an opt-out on the session constructor (or an option?) that would allow you to suppress these warnings. If we had that escape hatch, I don't think I'd be so opposed to always warning, though it still feels a little weird to me.